### PR TITLE
[GSoC] Add digit and text recongnition samples.

### PR DESCRIPTION
**Pull request by zihaomu**
Hi, this is my GSoC project.

I have closed the previous PR (https://github.com/opencv/opencv/pull/17462) which contains LeNet_digit recognition only.
And open a new one containing all code.

1. digit recognition

Take the live image from the camera, use connected component analysis to detect potential regions with each digit, and use the LeNet to classify.
With CPU only (i5-8300), it can achieve 12 FPS.
demo1

2. scene text detection and recognition
My laptop environment is CPU: i5-8300, GPU: 1050, Ubuntu 18
Take the live image from the camera, use EAST as text detector. After getting the detector output, crop these bounding box as the input of the text recognizer based on VGG. Finally, print the result near the box. Using GPU can achieve around 9FPS.
demo2
demo3

### Pull Request Readiness Checklist
See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under OpenCV (BSD) License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or other license that is incompatible with OpenCV
- [x] The PR is proposed to proper branch
- [x] There is reference to original bug report and related work
- [ ] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [ ] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

File: samples/dnn/text_detection_and_recognition.cpp [hunk 1]
```diff
@@ -0,0 +1,241 @@
+//  This example can not only detect text through the camera, but also perform real-time text recognition. Finally, display the results in the bounding box.
```
Comment (dkurt) [inline samples/dnn/text_detection_and_recognition.cpp:1]: What's the difference with https://github.com/opencv/opencv/blob/master/samples/dnn/text_detection.cpp?

File: samples/cpp/digits.cpp [hunk 2]
```diff
@@ -1,375 +1,180 @@
...
+//  This example provides a digital recognition based on LeNet-5 and connected component analysis.
+//  It makes it possible for OpenCV beginner to run dnn models in real time using only CPU.
+//  It can read pictures from the camera in real time to make predictions, and display the recognized digits as overlays on top of the original digits.
+//
+//  In order to achieve a better display effect, please write the number on white paper and occupy the entire camera.
+//
+//  You can follow the following guide to train LeNet-5 by yourself using the MNIST dataset.
+//  https://github.com/intel/caffe/blob/a3d5b022fe026e9092fc7abc7654b1162ab9940d/examples/mnist/readme.md
+//
+//  You can also download and train the model directly.
```
Comment (dkurt) [inline samples/cpp/digits.cpp:10]: Please consider comments from https://github.com/opencv/opencv/pull/17462

Comment (zihaomu) [reply samples/cpp/digits.cpp:10]: Sorry, it is my mistake.

File: samples/dnn/text_detection_and_recognition.cpp (hunk 1 above)
Comment (zihaomu) [reply samples/dnn/text_detection_and_recognition.cpp:1]: So far, the difference is the text recognition model. I use VGG, and text_detection.cpp uses CRNN.
And next, I will provide more text recognition model support (like ResNet+BLSTM, DensNet, and etc.) and better text detection methods.
Finally, I will analyze the performance of these models on OpenCV in terms of accuracy and speed.

Comment (dkurt) [reply samples/dnn/text_detection_and_recognition.cpp:1]: But text detection model is the same which leads to code duplication. Can you please add new text recognition model into this sample?

**Comment by alalek** [comment]
> I have closed the previous PR (#17462) which contains LeNet_digit recognition only.

Users may want to compare different algorithms/approaches on their tasks.

**Comment by zihaomu** [comment]
> > I have closed the previous PR (#17462) which contains LeNet_digit recognition only.
> 
> Users may want to compare different algorithms/approaches on their tasks.

Hi, I keep the original digits.cpp as digits_SVM.cpp, and add the proposal as digits_LeNet.cpp.

File: samples/dnn/text_detection_and_recognition.cpp (hunk 1 above)
Comment (zihaomu) [reply samples/dnn/text_detection_and_recognition.cpp]: Ok, I have added a link into text_detection.cpp.

File: samples/dnn/text_detection.cpp [hunk 3]
```diff
@@ -2,17 +2,19 @@
     Text detection model: https://github.com/argman/EAST
     Download link: https://www.dropbox.com/s/r2ingd0l3zt8hxs/frozen_east_text_detection.tar.gz?dl=1
 
-    Text recognition model taken from here: https://github.com/meijieru/crnn.pytorch
-    How to convert from pb to onnx:
-    Using classes from here: https://github.com/meijieru/crnn.pytorch/blob/master/models/crnn.py
```
Comment (dkurt) [inline samples/dnn/text_detection.cpp:7]: Don't remove existing usage - make sample work with both models.

Comment (zihaomu) [reply samples/dnn/text_detection.cpp:7]: The two text recognition preprocessing is different. Regarding the CRNN in https://github.com/meijieru/crnn.pytorch, I have provided a pre-trained model in my Google Drive. They have the same structure, but different pre-processing. In this case, how should I modify it?

Comment (dkurt) [reply samples/dnn/text_detection.cpp:7]: So, what's the difference between models? In terms of accuracy/efficiency?

Comment (zihaomu) [reply samples/dnn/text_detection.cpp:7]: Because the model structure is the same, the efficiency is the same.  About the accuracy, since more data for training, my pre-trained model can achieve 89.2% in IC13 dataset, while the original one is 86.7%.

Comment (dkurt) [reply samples/dnn/text_detection.cpp:7]: so they seems equal that means sample should support both of them. Try to manage it automatically or by flags.

Comment (zihaomu) [reply samples/dnn/text_detection.cpp:7]: Sorry for replying so late, I have added a flag to allow the sample to use both models at the same time.
In addition, the OCR model based on DenseNet+CTC and DenseNet+BiLSTM+CTC has been added https://drive.google.com/drive/folders/1cTbQ3nuZG-EKWak6emD_s8_hHXWz7lAr?usp=sharing.

File: samples/dnn/text_detection.cpp [hunk 4]
```diff
@@ -147,7 +164,7 @@ int main(int argc, char** argv)
 
                 std::string wordRecognized = "";
                 decodeText(result, wordRecognized);
-                putText(frame, wordRecognized, vertices[1], FONT_HERSHEY_SIMPLEX, 1.5, Scalar(0, 0, 255));
+                putText(frame, wordRecognized, vertices[1], FONT_HERSHEY_SIMPLEX, 1, Scalar(0, 0, 255));
```
Comment (zihaomu) [inline samples/dnn/text_detection.cpp:167]: The previous font is too big.

**Comment by vpisarev** [comment]
@zihaomu, thank you very much! I've tested your code and it works great!

Can you, please, also update text_detection.py to use the same model for detection and add OCR part?

File: samples/cpp/digits_LeNet.cpp [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+            // Extend the bounding box of connected component for easier recognition
+            if (stats.at<int>(i - 1, CC_STAT_AREA) > 80 && stats.at<int>(i - 1, CC_STAT_AREA) < 3000)
+            {
+                ifDrawingBox = true;
+                int left = stats.at<int>(i - 1, CC_STAT_HEIGHT) / 4;
+                getRectangle = Rect(stats.at<int>(i - 1, CC_STAT_LEFT) - left, stats.at<int>(i - 1, CC_STAT_TOP) - left, stats.at<int>(i - 1, CC_STAT_WIDTH) + 2 * left, stats.at<int>(i - 1, CC_STAT_HEIGHT) + 2 * left);
+                getRectangle &= basicRect;
+            }
+
+            if (ifDrawingBox)
```
Comment (vpisarev) [inline samples/cpp/digits_LeNet.cpp:134]: please, add after `ifDrawingBox`:

```
&& !getRectangle.empty()
```
otherwise it crashes almost immediately

**Comment by zihaomu** [comment]
> @zihaomu, thank you very much! I've tested your code and it works great!
> 
> Can you, please, also update text_detection.py to use the same model for detection and add OCR part?

@vpisarev Thank you for your reply.
Indeed, the implementation of text_detection.py does have errors. In order not to affect this PR of GSoC, I have created a new PR  https://github.com/opencv/opencv/pull/17992.

File: samples/dnn/text_detection.cpp [hunk 6]
```diff
@@ -138,7 +144,19 @@ int main(int argc, char** argv)
...
+                if(ocrFlag)  //Normalization in https://github.com/clovaai/deep-text-recognition-benchmark
+                {
+                    blobCrop = blobFromImage(cropped, 1.0/255.0, Size(), Scalar());
+                    blobCrop -= 0.5;
+                    blobCrop /= 0.5;
+                }
+                else  //Normalization in https://github.com/meijieru/crnn.pytorch
+                {
+                    blobCrop = blobFromImage(cropped, 1.0/127.5, Size(), Scalar::all(127.5));
+                }
```
Comment (dkurt) [inline samples/dnn/text_detection.cpp:158]: first case:
[0, 255] -> [0, 1] -> [-0.5, 0.5] -> [0, 1]
second:
[0, 255] -> [-127.5, 127.5] -> [0, 1]

They are equal

Comment (zihaomu) [reply samples/dnn/text_detection.cpp:158]: @dkurt Thank you. 
I just tested it, these two preprocessing is indeed the same, I will remove the ```ocrFlag``` in the code.

File: samples/dnn/text_detection.cpp [hunk 7]
```diff
@@ -2,10 +2,14 @@
...
     Download link: https://www.dropbox.com/s/r2ingd0l3zt8hxs/frozen_east_text_detection.tar.gz?dl=1
 
-    Text recognition model taken from here: https://github.com/meijieru/crnn.pytorch
+    CRNN Text recognition model taken from here: https://github.com/meijieru/crnn.pytorch
     How to convert from pb to onnx:
     Using classes from here: https://github.com/meijieru/crnn.pytorch/blob/master/models/crnn.py
 
+    More converted onnx text recognition models can be downloaded directly here:
+    Download link: https://drive.google.com/drive/folders/1cTbQ3nuZG-EKWak6emD_s8_hHXWz7lAr?usp=sharing
+    And these models taken from here:https://github.com/clovaai/deep-text-recognition-benchmark
```
Comment (dkurt) [inline samples/dnn/text_detection.cpp:11]: Model works with the same sample without any changes?

Comment (zihaomu) [reply samples/dnn/text_detection.cpp:11]: Yes, these two methods are fully compatible.

File: samples/cpp/digits_LeNet.cpp [hunk 8]
```diff
@@ -0,0 +1,180 @@
...
+}
+
+void predictor(Net net, const Mat &roi, int &classId, double &probability)
+{
+    Mat pred;
+    //Convert Mat to batch of images
+    Mat inputBlob = dnn::blobFromImage(roi, 1.0, Size(28, 28));
+    //set the network input
+    net.setInput(inputBlob);
+    //compute output, "prob" is the name of the output layer
```
Comment (dkurt) [inline samples/cpp/digits_LeNet.cpp:177]: please keep only `// compute output` (also add one whitespace after //)

Comment (zihaomu) [reply samples/cpp/digits_LeNet.cpp]: fixed

File: samples/dnn/text_detection.py [hunk 9]
```diff
@@ -1,11 +1,18 @@
...
+
+    More converted onnx text recognition models can be downloaded directly here:
+    Download link: https://drive.google.com/drive/folders/1cTbQ3nuZG-EKWak6emD_s8_hHXWz7lAr?usp=sharing
+    And these models taken from here:https://github.com/clovaai/deep-text-recognition-benchmark
+
     import torch
-    import models.crnn as CRNN
+    import models.crnn as crnn
+
     model = CRNN(32, 1, 37, 256)
```
Comment (dkurt) [inline samples/dnn/text_detection.py:16]: `import models.crnn as CRNN`
or
`model = crnn(32, 1, 37, 256)`

Comment (zihaomu) [reply samples/dnn/text_detection.py]: Thank you for pointing out this, fixed.

File: samples/cpp/digits_LeNet.cpp [hunk 10]
```diff
@@ -0,0 +1,180 @@
+//  This example provides a digital recognition based on LeNet-5 and connected component analysis.
```
Comment (dkurt) [inline samples/cpp/digits_LeNet.cpp:1]: As for me, it's better to name the files in lowercase:
```
digits_lenet
digits_svm
```

File: samples/dnn/text_detection.py [hunk 11]
```diff
@@ -1,12 +1,19 @@
...
     Using classes from here: https://github.com/meijieru/crnn.pytorch/blob/master/models/crnn.py
+
+    More converted onnx text recognition models can be downloaded directly here:
+    Download link: https://drive.google.com/drive/folders/1cTbQ3nuZG-EKWak6emD_s8_hHXWz7lAr?usp=sharing
+    And these models taken from here:https://github.com/clovaai/deep-text-recognition-benchmark
+
     import torch
-    import models.crnn as CRNN
-    model = CRNN(32, 1, 37, 256)
+    import models.crnn as crnn
```
Comment (dkurt) [inline samples/dnn/text_detection.py:14]: Can you please try to run this export script? It seems to me that should be
```python
from models.crnn import CRNN
model = CRNN(32, 1, 37, 256)
```

Comment (zihaomu) [reply samples/dnn/text_detection.py:14]: Solved
